mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        # gcloud config set core/project google.com:pm-hackathon
+
+        gce_num_instances = int(self._opts['gce_num_instances'])
+        assert gce_num_instances >= 2
+        bootstrap_cmds = [self._upload_mgr.uri(self._master_bootstrap_script_path)]
+
+        # Change machine type
+        zone_uri = 'https://www.googleapis.com/compute/v1/projects/%(project)s/zones/%(zone)s' % dict(project=self._gcp_project, zone=self._gcp_zone)
+        machine_uri = "%(zone_uri)s/machineTypes/%(machine_type)s" % dict(zone_uri=zone_uri, machine_type=self._opts['gce_machine_type'])
+
+        # Task tracker
+        master_instance_group_config = {
+            "numInstances": 1,
+            "machineTypeUri": machine_uri,
+            "isPreemptible": False,
+        }
+
+        # Compute + storage
+        worker_instance_group_config = {
+            "numInstances": gce_num_instances,
+            "machineTypeUri": machine_uri,
+            "isPreemptible": False,
+        }
+
+        # TODO - mtai @ davidmarin - Add support for secondary worker configs
+        # Compute ONLY
+        secondary_worker_instance_group_config = {
+            "numInstances": 0,
+            "machineTypeUri": machine_uri,
+            "isPreemptible": True,
+        }
+
+        cluster_data = {
+            'projectId': self._gcp_project,
+            'clusterName': self._gcp_cluster,
+            'config': {
+                'gceClusterConfig': {
+                    'zoneUri': zone_uri
+                },
+                'masterConfig': master_instance_group_config,
+                'workerConfig': worker_instance_group_config,
+                # 'secondaryWorkerConfig': secondary_worker_instance_group_config,
+                'initializationActions': [
+                    dict(executableFile=bootstrap_cmd) for bootstrap_cmd in bootstrap_cmds
+                ]
+            }
+        }
+
+        self.dataproc_client.clusters().create(
+            projectId=self._gcp_project,
+            region=self._gcp_region,
+            body=cluster_data).execute()
+        cluster_state = None
+
+        # TODO - mtai @ davidmarin - See https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.clusters#State
+        # TODO - mtai @ davidmarin - We should probably pull these strings to constants
+
+        # TODO - mtai @ davidmarin - Can end up in infinite loop if cluster_state doesn't end up in (RUNNING,ERROR)
+        while bool(cluster_state != 'RUNNING'):
+            result_describe =  self.dataproc_client.clusters().get(projectId=self._gcp_project,
+                                               region=self._gcp_region,
+                                               clusterName=self._gcp_cluster).execute()
+    
+            cluster_state = result_describe['status']['state']
+            if cluster_state == 'ERROR':
+                raise Exception(result_describe)
+    
+            log.info('Waiting for cluster to launch... sleeping 5 second(s)')
```

Comment:
Once everything else is working, let's think about how often we want to print status messages to the user and what they should look like (just cleaned this up in EMR).